### PR TITLE
Feature/swizzle app delegate

### PR DESCRIPTION
**Pull request by clpolet**
Forward app delegate calls to the SDK

* Add MSAppDelegateForwarder responsible for setting up and managing app delegate forwarding
* Swizzle methods from the app delegate to forward them to SDK modules
* Swizzle only delegate methods required by modules, modules will have to register them
* Add MSCustomAppDelegate protocol that modules conform to in order to receive forwarded messages
* Several module can subscribe to the same delegate method
* Results of the delegate method is chained through all the delegate implementations starting from the original one
* Add unit tests to cover most of the scenarios
* Add a mocked app delegate for easy of use in tests

### CODE REVIEW

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  }
+}
+
+#pragma mark - Swizzling
+
++ (void)swizzleSelector:(SEL)selector {
+  NSString *oldOriginalSelector = NSStringFromSelector(selector);
+  NSString *newOriginalSelector;
+
+  // Don't swizzle twice the same selector.
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: move `twice` at the end of the comment.

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      if (methodAdded || ![delegate respondsToSelector:selector]) {
+        class_replaceMethod(appDelegateClass, selector, method_getImplementation(customMethod),
+                            method_getTypeEncoding(customMethod));
+        [self.swizzledSelectors addObject:oldOriginalSelector];
+        MSLogDebug([MSMobileCenter logTag], @"App delegate selector %@ swizzled.", oldOriginalSelector);
+      }
+
+      // The new method may not be added (i.e.: in case of conflicts), should be rare as it's prefixed.
+      else {
+        MSLogError([MSMobileCenter logTag],
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: good catch of the conflict case!

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.m [hunk 3]
```diff
@@ -28,6 +24,15 @@ + (UIApplication *)sharedApp {
                                                                                                 sharedAppSel);
 }
 
++ (id<UIApplicationDelegate>)sharedAppDelegate {
+  return [self sharedApp].delegate;
+}
+
+
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.m]: unnecessary line.

File: MobileCenter/MobileCenter/MSMobileCenter.m [hunk 4]
```diff
@@ -11,9 +12,7 @@
 static MSMobileCenter *sharedInstance = nil;
 static dispatch_once_t onceToken;
 
-/**
- * Base URL for HTTP Ingestion backend API calls.
- */
+// Base URL for HTTP Ingestion backend API calls.
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSMobileCenter.m]: This is a doc. Please revert. We use // for single line and /* */ for multiple lines but /** */ is for docs.

Comment (clpolet) [inline MobileCenter/MobileCenter/MSMobileCenter.m]: Then we should do it everywhere, it's not consistent.

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * Enable/Disable Application forwarding.
+ */
+@property(nonatomic, class) BOOL enabled;
+
+/**
+ * Add a delegate.
+ *
+ * @param delegate A delegate.
+ */
++ (void)addDelegate:(id<MSAppDelegate>)delegate;
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.h]: Header docs should mention that it's a thread safe method.

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+  BOOL methodAdded = NO;
+  
+  // Replace original implementation by the custom one.
+  if (originalMethod){
+    originalImp = method_setImplementation(originalMethod, customImp);
+  }
+  
+  /*
+   * The original class may not implement the selector (e.g.: optional method from protocol), 
+   * add the method to the original class and associate it to the custom implementation.
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: associate it `with` the custom implementation

Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: In addition, I find this hard to read as the `else` comes after the comment. Maybe you can move the comment into the `else` block?

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    Method customMethod = class_getInstanceMethod(self, customSelector);
+    methodAdded = class_addMethod(originalClass, originalSelector, customImp, method_getTypeEncoding(customMethod));
+  }
+  
+  // Validate swizzling.
+  if (!originalImp && !methodAdded) {
+    MSLogError([MSMobileCenter logTag],
+                @"Cannot swizzle selector %@ of class %@. You will have to explicitly call APIs from "
+                @"Mobile Center in your app delegate implementation.", originalSelectorString, originalClass);
+  }else{
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: new line
```
}
else {
...
}```

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+  return originalImp;
+}
+
++ (void)addAppDelegateSelectorToSwizzle:(SEL)selector {
+  [self.selectorsToSwizzle addObject:NSStringFromSelector(selector)];
+}
+
+#pragma mark - UIApplication
+
+
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: Remove 1 newline.

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+/*
+ * The application starts querying its delegate for its implementation as soon as it is set then may never query again.
+ * It means that if the application delegate doesn't implement an optional method of the `UIApplicationDelegate` protocol 
+ * at that time then that method may never be called even if added later via swizzling. This is why the application
+ * delegate swizzling should happen at the time it is set to the application object.
+ */
+
+@implementation UIApplication (MSSwizzling)
+
++ (void)load{
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: Code formatting is odd, I think. There should be a space before the `{`.

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m (hunk 7 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: I don't think we have this rule, should we?

Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: I think that would probably be another task.

File: MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m (hunk 9 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m]: True I'll run the formatting this file, I probably missed it.

File: MobileCenterDistribute/MobileCenterDistribute/Internals/MSDistributeAppDelegate.m [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  return [self openURL:url returnedValue:returnedValue];
+}
+
+-(BOOL)application:(__attribute__((unused)) UIApplication *)app openURL:(NSURL *)url options:(__attribute__((unused)) NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options returnedValue:(BOOL)returnedValue{
+  return [self openURL:url returnedValue:returnedValue];
+}
+
+#pragma mark - Private
+
+-(BOOL)openURL:(NSURL *)url returnedValue:(BOOL)returnedValue{
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/Internals/MSDistributeAppDelegate.m]: Can you run reformat one more time? Make sure you run formatter against all source files in the PR.

**Comment by clpolet** [comment]
@jaelim-ms I though that's what I did, do you have a file in particular?

File: MobileCenter/MobileCenter.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -730,7 +743,7 @@
 				3844FF221E8C2716003E9194 /* MSAbstractLog.h in Headers */,
 				3844FF151E8C253F003E9194 /* MSLogWithPropertiesInternal.h in Headers */,
 				F803BBF31E8E3677004B1E7A /* MSCustomProperties.h in Headers */,
-				B2C3C1921DB9864600CB83F7 /* MSWrapperSdk.h in Headers */,
+				6E04016A1D1C9E1F0051BCFA /* MSMobileCenter.h in Headers */,
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter.xcodeproj/project.pbxproj]: Is this just a diff issue?

Comment (clpolet) [inline MobileCenter/MobileCenter.xcodeproj/project.pbxproj]: Yes, you won't see this one in the global change.

**Pull request merged**